mrmustard/math/lattice/strategies/vanilla.py

```diff
@@ -147,26 +147,44 @@ def vanilla_vjp(G, c, dLdG) -> tuple[ComplexMatrix, ComplexVector, complex]:  #
     Returns:
         tuple[np.ndarray, np.ndarray, complex]: dL/dA, dL/db, dL/dc
     """
-    D = G.ndim
+    shape = G.shape
+
+    # calculate the strides
+    strides = shape_to_strides(np.array(shape))
+
+    # linearize G
+    G_lin = G.flatten()
 
     # init gradients
+    D = G.ndim
     dA = np.zeros((D, D), dtype=np.complex128)  # component of dL/dA
     db = np.zeros(D, dtype=np.complex128)  # component of dL/db
     dLdA = np.zeros_like(dA)
     dLdb = np.zeros_like(db)
 
-    # initialize path iterator
-    path = np.ndindex(G.shape)
+    # initialize the indeces.
+    # ``index`` is the index of the flattened output tensor, while
+    # ``index_u_iter`` iterates through the unravelled counterparts of
+    # ``index``.
+    index = 0
+    index_u_iter = np.ndindex(shape)
+    next(index_u_iter)
 
-    # skip first index
-    next(path)
+    for index_u in index_u_iter:
+        index += 1
 
-    # iterate over the rest of the indices
-    for index in path:
-        dA, db = steps.vanilla_step_grad(G, index, dA, db)
-        dLdA += dA * dLdG[index]
-        dLdb += db * dLdG[index]
+        ns = lower_neighbors(index, strides, 0)
+
+        for i, _ in enumerate(db):
+            _, n = next(ns)
+            db[i] = np.sqrt(index_u[i]) * G_lin[n]
+            dA[i, i] = 0.5 * np.sqrt(index_u[i] * (index_u[i] - 1)) * G_lin[n - strides[i]]
+            for j in range(i + 1, len(db)):
+                dA[i, j] = np.sqrt(index_u[i] * index_u[j]) * G_lin[n - strides[j]]
+
+        dLdA += dA * dLdG[index_u]
+        dLdb += db * dLdG[index_u]
 
-    dLdc = np.sum(G * dLdG) / c
+    dLdc = np.sum(G_lin.reshape(shape) * dLdG) / c
```

Comment:
Is this faster/slower/same?
```suggestion
    dLdc = np.sum(G_lin * dLdG.flat).reshape(shape)
```